Update ghost particles without rebuilding them

blaisb opened the pull request:
This is a work in progress ([WIP])
This PR aims at allowing for a function that updates the ghost particles without rebuilding them from scratch. Many numerical methods don't require that the particle interact with the mesh at every time step (e.g. MD based methods or DEM). This PR aims at addressing #11093 .

- [X] Create an update_ghost_particles function and move the necessary containers to particle_handler class member. In the update_ghost_particles remove the loop to generate the ghost_particles_by_domain
- [x] Add a unit test for ghost updates
- [ ] Modify send_rcv_particles (or write a new function alltogether) to enable updating existing particles without recreating them from scratch. This part i will do with the advices @peterrum (thanks again for helping me with this :).

The first modification already implemented allowed us a speed up of 100% in our DEM code on 2 cores (from 280s to 127s) and a 30% (103s to 72s) at 16 core. So this is already extremely significant. I think it's the last optimization we need in our case. This will also make the particle_handler a very very flexible class!

peterrum commented:
> The second issue is more profound. At every call, exchange_ghost_particles() clears the entire ghost_particles structure and rebuilds it from scratch, assuming that ghost particles may have changed cells and everything. The underlying assumption is that we are calling sort_particles_into_subdomain_and_cells at every time step. Regretfully, this is not what we are doing.

@blaisb Could you quickly explain me the reason for why you don't have to call `sort_particles_into_subdomain_and_cells` in the case of DEM every time step. Maybe this is a stupid question and I am not aware about the differences of DEM and PIC.

blaisb commented:
> > The second issue is more profound. At every call, exchange_ghost_particles() clears the entire ghost_particles structure and rebuilds it from scratch, assuming that ghost particles may have changed cells and everything. The underlying assumption is that we are calling sort_particles_into_subdomain_and_cells at every time step. Regretfully, this is not what we are doing.
> 
> @blaisb Could you quickly explain me the reason for why you don't have to call `sort_particles_into_subdomain_and_cells` in the case of DEM every time step. Maybe this is a stupid question and I am not aware about the differences of DEM and PIC.

In DEM  the particle interact between themselves every timestep (we do particle-particle force calculation through either contact models or non-contact models (e.g. an LJ potential)). However, we want the particles to interact with the mesh every **n** iterations, where **n** is a number between 1 and 1000. Consequently, when we calculate particle-particle interaction, we don't need to have the particles exactly sorted into the subdomain (we don't need their reference location). The only thing we need is their properties (velocity, diameter) and their physical location. This is sufficient to do the particle-particle interaction. Thus, we need a mechanism to update the properties and the physical location of the ghost particles.

Right now we are using exchange_ghost to achieve this. But exchange ghost has a very large overhead. Right now, a single core DEM benchmark simulation in Lethe runs in 120s on one core, but runs in 287s on two cores because of the overhead of exchange_ghost_particles. The present PR in it's current state brings the two-core cost to 120s, removing 160s of overhead just because of the loop that is removed when calling update_ghosts every iteration and exchange ghost every **n** iterations.

I know this is usage is not what was originally intended for particles. However, I think providing this update mechanism does not break the code for anyone. It provides a much more flexible implementation that allow a code to do particle-particle stuff at one time scale and particle-mesh stuff at another time scale. Our ultimate goal is to be able to build the first families of DEM scheme where DEM can be coupled with a CFD solver without requiring that we couple two existing code (e.g. OpenFOAM and LIGGGHTS, Fluent and EDEM, etc.). We think this will provide at least a 10x speed-up over the current state of the art just by using existing deal.II facilities and what we added in Lethe.

I hope that clearly explains it :)!.

peterrum commented:
Thanks for the explanation @blaisb! If I see it correctly you make the simplified assumption that no new ghost-cell particles are created during the n time steps, which is probably acceptable in your case. We can discuss this later offline.

peterrum commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -707,6 +707,13 @@ namespace Particles
     exchange_ghost_particles();
 
+    /**
+     * Update all particles that live in cells that are ghost cells to
```
Could you be more explicit what *update* means in this context.

peterrum commented on source/particles/particle_handler.cc:
```diff
@@ -1222,6 +1225,31 @@ namespace Particles
+  ParticleHandler<dim, spacedim>::update_ghost_particles()
+  {
+    // Nothing to do in serial computations
+    const auto parallel_triangulation =
+      dynamic_cast<const parallel::TriangulationBase<dim, spacedim> *>(
+        &*triangulation);
```
```suggestion
    const auto parallel_triangulation =
      dynamic_cast<const parallel::TriangulationBase<dim, spacedim> *>(
        triangulation->get());
```

blaisb replied:
This does not seem to work. No member get in dealii::Triangulation<1,1> ?

peterrum replied:
Oh, triangulation is *SmartPoint*. Just ignore my comments ;)

peterrum commented on source/particles/particle_handler.cc:
```diff
@@ -1222,6 +1225,31 @@ namespace Particles
+    const auto parallel_triangulation =
+      dynamic_cast<const parallel::TriangulationBase<dim, spacedim> *>(
+        &*triangulation);
+    if (parallel_triangulation != nullptr)
+      {
+        if (dealii::Utilities::MPI::n_mpi_processes(
+              parallel_triangulation->get_communicator()) == 1)
+          return;
+      }
+    else
+      return;
```
```suggestion
    if (parallel_triangulation == nullptr || 
        Utilities::MPI::n_mpi_processes(parallel_triangulation->get_communicator()) == 1)
      return;
```

peterrum commented on source/particles/particle_handler.cc:
```diff
@@ -1679,8 +1707,8 @@ namespace Particles
 
     switch (status)
       {
-        case parallel::distributed::Triangulation<dim, spacedim>::CELL_PERSIST:
-          {
+          case parallel::distributed::Triangulation<dim,
+                                                    spacedim>::CELL_PERSIST: {
```
I guess this is a formatting issue.

peterrum commented on source/particles/particle_handler.cc:
```diff
@@ -1222,6 +1225,31 @@ namespace Particles
+    else
+      return;
+
+#ifdef DEAL_II_WITH_MPI
```
Is this guard actually needed?

peterrum reviewed (commented):
Looks good to me. Once you have added a test, I think this PR is ready to be merged.

The point:
> Modify send_rcv_particles (or write a new function alltogether) to enable updating existing particles without recreating them from scratch.

could be tackled in a follow-up PR. What do you think?

blaisb commented:
This last push finishes task 2 of this PR.
I added a unit test which has a property and a position, which is changed and the ghosts are updated. It is simple, but it really does the job. I also applied @peterrum  comments and fixed indentation.

peterrum commented:
/rebuild

peterrum reviewed (approved):
Thanks!

blaisb commented:
> Thanks!

No problem. If it is good for you, I suggest we go as you suggested. We could merge this PR and then together work on the third (and last) item. :)!

peterrum commented:
> third (and last) item. :)!

Third but definitely not the last item ;)
